composer/core/time.py

```diff
@@ -356,7 +356,7 @@ def from_timestring(cls, timestring: str) -> Time:
         return cls(value, unit)
 
 
-class Timer(Serializable):
+class Timestamp(Serializable):
```

Comment:
While we originally envisioned that the `Timer` class would use a `Timestamp` internally, I do appreciate how binding the latest `Timestamp` to the state, rather than binding the `Timer` to the state, simplifies the logic wherever the timer is used (i.e. no need to call `.get_timestamp()` anymore). Thanks for doing it this way!

One request: Since the `Timestamp` is supposed to be treated like an immutable object (i.e. once it is created, it is never modified), would it be possible to remove the methods `on_epoch_complete` and `on_batch_complete` from the timestamp, and instead hard-code this logic directly into the Trainer? The trainer can create a new Timestamp object, and bind the new object to the State, instead of modifying the existing Timestamp object.


For example, when the epoch is complete, the trainer can set `state.timestamp = Timestamp(epoch=state.timestamp.epoch + Time(1, TimeUnit.EPOCH), batch=state.timestamp.batch, sample=state.timestamp.sample, token=state.timestamp.token, batch_in_epoch=Time(0, TimeUnit.BATCH), ...).` The logic would be parallel for where `on_batch_complete` is currently called.

Reply:
Thank you so much for the comments and suggestions. Very reasonable and well put. Sorry had a busy week, haven't got time to resume the work, I will try and get address your comments and update the PR soon. Thanks again!

Reply:
Since these methods were used throughout tests, and could be useful outside of the Mosaic trainer, I kept the logic of `on_batch_complete` and `on_epoch_complete` as part of the `Timestamp`, but renamed these methods to `to_next_batch` and `to_next_epoch`, respectively, and have them return a copy of the `Timestamp` rather than keep it in place.